examples/gitlab-asg/Makefile

```diff
@@ -1,44 +1,77 @@
-.PHONY: render-tls-configs render-tfvars render-ssh-config generate-ssh-key generate-tfvars generate-tls-certs upload-tls install-cfssl network plan apply test destroy
+.PHONY: all shell render-tls-configs generate-ssh-key generate-tls-certs install-cfssl upload-tls-cert delete-tls-cert render-tfvars network plan apply render-ssh-config init-ebs-volume create-activate-swap terminate-gitlab-server register-gitlab-runner destroy clean help
 
 .DEFAULT_GOAL = help
 
-## Use envsubst and vars.env to render the json config files for cfssl
+include vars.mk
+
+PATH:=${PWD}/bin:${PATH}
+
+GNUMAKEFLAGS := --no-print-directory
+
+## Do everything
+all:
+	# Use recursive makes instead of dependencies to show progress
+	# If something breaks, we can see which target it was in
+	$(MAKE) install-cfssl
+	$(MAKE) render-tls-configs
+	$(MAKE) generate-tls-certs
+	$(MAKE) upload-tls-cert
+	$(MAKE) generate-ssh-key
+	$(MAKE) render-tfvars
+	$(MAKE) network
+	$(MAKE) plan
+	$(MAKE) apply
+	$(MAKE) render-ssh-config
+	@echo "Waiting for instance to initialize..."
+	sleep 240
+	$(MAKE) init-ebs-volume
+	$(MAKE) terminate-gitlab-server
+	@echo "Waiting for instance to reinitialize..."
+	sleep 240
+	$(MAKE) render-ssh-config
+
+## Run an interactive shell with all environment variables defined
+shell:
+	@bash
+
+## Use envsubst to render the json config files for cfssl
 render-tls-configs:
-	@. ./vars.env && envsubst <ca-csr.tpl > ca-csr.json
-	@. ./vars.env && envsubst <cert-csr.tpl > cert-csr.json
+	@envsubst <ca-csr.tpl > ca-csr.json
+	@envsubst <cert-csr.tpl > cert-csr.json
 
 # Use ssh-keygen to generate a keypair for SSH
 generate-ssh-key:
-	@ssh-keygen -q -N "" -b 4096 -C "test" -f ./id_rsa
+	@ssh-keygen -q -N "" -b 4096 -C "test" -f id_rsa
 	
 ## Use cfssl to generate a CA and self-signed server SSL certificate for AWS ELB
 generate-tls-certs:
-	@./cfssl gencert -initca ca-csr.json | cfssljson -bare ca -
-	@./cfssl gencert -ca=ca.pem -ca-key=ca-key.pem -config=cert-config.json -profile=gitlab cert-csr.json | cfssljson -bare gitlab
+	@cfssl gencert -initca ca-csr.json | cfssljson -bare ca -
+	@cfssl gencert -ca=ca.pem -ca-key=ca-key.pem -config=cert-config.json -profile=gitlab cert-csr.json | cfssljson -bare gitlab
 	@ls -Alh
 	@echo "ca-key.pem is the file to keep very safe"
 
 ## use curl to download and install the cfssl executable
 install-cfssl:
-	@curl -s -L -o ./cfssl https://pkg.cfssl.org/R1.2/cfssl_linux-amd64
-	@curl -s -L -o ./cfssljson https://pkg.cfssl.org/R1.2/cfssljson_linux-amd64
-	@chmod +x ./cfssl ./cfssljson
-	@./cfssl version
+	@mkdir -p bin
+	@curl -s -L -o bin/cfssl https://pkg.cfssl.org/R1.2/cfssl_linux-amd64
```

Comment:
why use `./bin` over `~/bin` ?

Reply:
I wanted to keep everything local and not touch the user's global config at all. So, in particular, if there's already a `cfssl` in `~/bin` we shouldn't be stomping on it. We remove `./bin` at the end, in `clean`, so it should be a case of "We Were Never Here".

Also, I think the default should be to install `cfssl` from a package, so this should be used only by someone who's in a hurry and just wants to try things out quickly, which is another reason we shouldn't be leaving stuff behind in the global env.

I did consider making the download conditional on whether we can detect `cfssl` already being on the `PATH` somewhere (eg using `which` or `type`).